### ISSUE TITLE
Support write operation 

### ISSUE THREAD

**Issue by quentingodeau** [duckdb/duckdb_azure#44]
Hi,

It's not really an issue but more an insight on what I plan to work on.
For the moment I didn't start but when I do I will post here a message to notify. If someone start before me please let me know ;)

**Comment by samansmink** [comment]
Thats great to hear :) Let me know if you need any help, feel free to reach out to me on the duckdb discord or sam@duckdblabs.com

**Comment by csubhodeep** [comment]
Hello ducdkb, team nice to see that this has already been posted as an issue. I and my team would also love to have this as a feature. Just to add some context here - we are working with ETL pipelines in my company that mostly use `pandas` however for some performance related reasons we have started migrating to `duckdb` and therefore, while we could already enjoy the "parquet import" feature we would like to have a parquet export capability like in [S3](https://duckdb.org/docs/guides/import/s3_export).

Just a small example,
```sql
CREATE TABLE weather (
    city    VARCHAR,
    temp_lo INTEGER, -- minimum temperature on a day
    temp_hi INTEGER, -- maximum temperature on a day
    prcp    REAL,
    date    DATE
);

INSERT INTO weather VALUES ('San Francisco', 46, 50, 0.25, '1994-11-27');
INSERT INTO weather VALUES ('San Francisco', 46, 50, 0.25, '1994-11-27');
INSERT INTO weather VALUES ('San Francisco', 46, 50, 0.25, '1994-11-27');

COPY weather TO 'az://⟨my_container⟩/⟨my_file⟩.⟨parquet_or_csv⟩'
```

**Comment by samansmink** [comment]
@csubhodeep you can try using [fsspec](https://duckdb.org/docs/guides/python/filesystems) if you're on python, they should have azure support

**Comment by csubhodeep** [comment]
> @csubhodeep you can try using [fsspec](https://duckdb.org/docs/guides/python/filesystems) if you're on python, they should have azure support

ok thanks a lot. I will try it.

**Comment by csubhodeep** [comment]
Thanks again! I tried to use the `fsspec` library in conjunction with the [`adlfs`](https://github.com/fsspec/adlfs). TLDR - NO success

Here are what I tried:
```shell
>>> storage_account_name = "our_account"
>>> container_name = "our_container"
>>> account_creds = <our_key>
>>> duckdb.register_filesystem(filesystem('abfs', connection_string=account_creds))

>>> duckdb.sql("CREATE OR REPLACE TABLE test_table (a INTEGER, b VARCHAR(100))")
>>> duckdb.sql("INSERT INTO test_table VALUES (1, 'a'), (2, 'b'), (3, 'c')")
>>> duckdb.sql("SELECT * FROM test_table")
┌───────┬─────────┐
│   a   │    b    │
│ int32 │ varchar │
├───────┼─────────┤
│     1 │ a       │
│     2 │ b       │
│     3 │ c       │
└───────┴─────────┘
>>> write_query = f"COPY test_table TO 'https://{storage_account_name}.blob.core.windows.net/{container_name}/test.parquet' (FORMAT 'parquet')"
---------------------------------------------------------------------------
IOException                               Traceback (most recent call last)
Cell In[41], [line 2](vscode-notebook-cell:?execution_count=41&line=2)
      [1](vscode-notebook-cell:?execution_count=41&line=1) # dump it as parquet
----> [2](vscode-notebook-cell:?execution_count=41&line=2) duckdb.sql(write_query)

IOException: IO Error: Cannot open file "https://<storage_account_name>.blob.core.windows.net/<container_name>/test.parquet": No such file or directory
>>> write_query = f"COPY test_table TO 'az://{storage_account_name}.blob.core.windows.net/{container_name}/test.parquet' (FORMAT 'parquet')"
---------------------------------------------------------------------------
NotImplementedException                   Traceback (most recent call last)
Cell In[43], [line 2](vscode-notebook-cell:?execution_count=43&line=2)
      [1](vscode-notebook-cell:?execution_count=43&line=1) # dump it as parquet
----> [2](vscode-notebook-cell:?execution_count=43&line=2) duckdb.sql(write_query)

NotImplementedException: Not implemented Error: Writing to Azure containers is currently not supported
```
Please let me know if I am doing something wrong.

**Comment by samansmink** [comment]
could you try the `abfs://` urls instead?  the `az://` once are triggering our auto-installation routing the requests through the azure extension. Alternatively, disable autoloading with `set autoinstall_known_extensions=false;`

**Comment by csubhodeep** [comment]
After trying the suggestion above, here are the results:
```shell
Exception ignored in: <function AzureBlobFile.__del__ at 0x7feb3d5a4280>
Traceback (most recent call last):
  File "/workspaces/rev_man_sys/venv/lib/python3.8/site-packages/adlfs/spec.py", line 2166, in __del__
    self.close()
  File "/workspaces/rev_man_sys/venv/lib/python3.8/site-packages/adlfs/spec.py", line 1983, in close
    super().close()
  File "/workspaces/rev_man_sys/venv/lib/python3.8/site-packages/fsspec/spec.py", line 1932, in close
    self.flush(force=True)
  File "/workspaces/rev_man_sys/venv/lib/python3.8/site-packages/fsspec/spec.py", line 1803, in flush
    if self._upload_chunk(final=force) is not False:
  File "/workspaces/rev_man_sys/venv/lib/python3.8/site-packages/fsspec/asyn.py", line 118, in wrapper
    return sync(self.loop, func, *args, **kwargs)
  File "/workspaces/rev_man_sys/venv/lib/python3.8/site-packages/fsspec/asyn.py", line 103, in sync
    raise return_result
  File "/workspaces/rev_man_sys/venv/lib/python3.8/site-packages/fsspec/asyn.py", line 56, in _runner
    result[0] = await coro
  File "/workspaces/rev_man_sys/venv/lib/python3.8/site-packages/adlfs/spec.py", line 2147, in _async_upload_chunk
    raise RuntimeError(f"Failed to upload block: {e}!") from e
RuntimeError: Failed to upload block: The specifed resource name contains invalid characters.
RequestId:3545381a-d01e-0083-2346-6efd88000000
Time:2024-03-04T15:11:28.7881722Z
ErrorCode:InvalidResourceName
Content: <?xml version="1.0" encoding="utf-8"?><Error><Code>InvalidResourceName</Code><Message>The specifed resource name contains invalid characters.
RequestId:3545381a-d01e-0083-2346-6efd88000000
Time:2024-03-04T15:11:28.7881722Z</Message></Error>!
```
Is it more of an `adlfs` issue?

**Comment by samansmink** [comment]
I think you're not setting the connection string correctly, you're setting it to your key it appears.

Let's move this discussion elsewhere though as this is no longer about this issues. Please check if you're actually using fsspec correctly. If things are still wrong and it appears to be duckdb side, feel free to open an issue in duckdb/duckdb

**Comment by csubhodeep** [comment]
I have managed to make it work the issue at my end was this part of the path - `{storage_account_name}.blob.core.windows.net` which even the `adlfs` library does not like for some reason. Not using that in the path just works fine. Thanks a lot again for all the guidance.

**Comment by patialashahi31** [comment]
Hi. I guess I am facing the same issue. I am using kotlin. Is there any workaround for this?

Thanks

**Comment by quentingodeau** [comment]
Hello,

For kotlin I'm not aware of a workaround :(
Nevertheless I have start to work on the issue but I think I will not be able to make it work until the following [PR](https://github.com/duckdb/duckdb/pull/11259) is merge

**Comment by bayees** [comment]
+1 for this feature

**Comment by IlijaStankovski** [comment]
Did the write operation feature via duckdb extensions end up being merged in the 1.0.0 release? I am currently using the 1.0.0 release and write operations via the azure extension still fails when using the duckdb node.js libraries. I am using both az:// and abfss:// when trying to write back to a parquet file hosted in Azure Storage Account (az://) or Azure Data Lake (abfss://).

Write operation error message:

Error: Not implemented Error: AzureDfsStorageFileSystem: FileExists is not implemented! {errno: -1, code: 'DUCKDB_NODEJS_ERROR', errorType: 'Not implemented', stack: 'Error: Not implemented Error: AzureDfsStorageFile

Any clarification on the write operation support state to Azure via node.js is highly appreciated.

Thank you.

**Comment by samansmink** [comment]
@IlijaStankovski It has not. I would like to pick this up at some point but I can't give a timeline here unfortunately.

**Comment by miquelmartorell** [comment]
+1 I support this request. It is really necessary. Thank you

**Comment by djouallah** [comment]
it will be nice to have :)

**Comment by davidsteinar** [comment]
@samansmink any updates on this?

**Comment by samansmink** [comment]
@davidsteinar not yet, sorry! I agree this is one the more high prio features we should look into but i can not give a timeline yet

**Comment by davidsteinar** [comment]
@samansmink for sure, it would enable me to throw away our existing data warehouse completely! Do you have any idea on the size/complexity of implementing this?

**Comment by nixent** [comment]
+1 to support the feature

**Comment by pantonis** [comment]
I don't think duckdb has azure on top of its list. don't know why though. Azure gen2 is absolutely the top storage service for now.. Hope someday somebody implements that.

**Comment by DrewScatterday** [comment]
I'm trying to do the workaround @csubhodeep did and I'm having some issues. I have an azure storage account `dscatterday` with a blob container `main`. I have env variables for connection string that has the format:
`DefaultEndpointsProtocol=https;AccountName=dscatterday;AccountKey=asdadasdad==;EndpointSuffix=core.windows.net`

 Here is a simple python code example to replicate my issue: 

```python
import duckdb
import os
from fsspec import filesystem

duckdb.register_filesystem(filesystem("abfs", connection_string=os.getenv("AZURE_CONN_STR")))
con = duckdb.connect()
con.execute("CREATE TABLE parqdata AS SELECT * FROM read_parquet('C:\\Users\\dre11620\\Downloads\\NY\\smallNYTaxi.parquet\\*.parquet', hive_partitioning = true);")
con.execute(f'''COPY parqdata TO 'abfs://main/test.parquet' (FORMAT PARQUET, PER_THREAD_OUTPUT);''') 
```

But I keep getting back a duckdb error:
```
duckdb.duckdb.IOException: IO Error: Failed to create directory "abfs://main/test.parquet": The filename, directory name, or volume label syntax is incorrect.
```
What am I missing?

**Comment by DrewScatterday** [comment]
Figured out the issue, I needed to do `register_filesystem` on the db connection object. Works now!

**Comment by dlaplante75** [comment]
+1 vote for this feature

**Comment by natemcintosh** [comment]
@DrewScatterday is the connection string the storage account URL? And if not, how can I find it for a given storage container?

**Comment by jordandakota** [comment]
Fsspec works from the python side and supports the data warehouse etl workflow, but with new users connecting with ducklake, we'll want this supported in the extension and via the duckdb web or dbeaver ide or similar connection methods. This issue just gained a lot of demand from the community.

**Comment by crazy-treyn** [comment]
> Fsspec works from the python side and supports the data warehouse etl workflow, but with new users connecting with ducklake, we'll want this supported in the extension and via the duckdb web or dbeaver ide or similar connection methods. This issue just gained a lot of demand from the community.

+1, that's why I'm here, now blocked on using DuckLake with Azure..

**Comment by csubhodeep** [comment]
+1 Following up on this one, would be really helpful if we could make this work without `fsspec`. Thanks !

**Comment by crazy-treyn** [comment]
> Fsspec works from the python side and supports the data warehouse etl workflow, but with new users connecting with ducklake, we'll want this supported in the extension and via the duckdb web or dbeaver ide or similar connection methods. This issue just gained a lot of demand from the community.

@jordandakota  Would you be able to share how you were able to get ducklake working with ADLS Gens/Blob storage using fsspec? I did what you mentioned above, and now getting a different error when I try to create a table in ducklake using an already existing container `data` in my ADLS Gen2 storage account: 

`IO Error: Failed to create directory "abfs://data/": No such file or directory`

Code: 
```python
import os
import duckdb
import adlfs
from fsspec import filesystem

con = duckdb.connect()
duckdb.register_filesystem(filesystem("abfs", connection_string=os.getenv("AZURE_CONNECTION_STRING")))

con.execute("""
    INSTALL ducklake;
    INSTALL postgres;

    LOAD ducklake;
    LOAD postgres;
""")

con.execute(f"""
    CREATE SECRET (
        TYPE postgres,
        HOST '{os.environ["POSTGRES_SERVER"]}',
        PORT 5432,
        DATABASE postgres,
        USER '{os.environ["POSTGRES_USER"]}',
        PASSWORD '{os.environ["POSTGRES_PASS"]}'
    );
""")

# Setup data path to container "data" that already exists in my ADLS Gen2 Storage Account
con.execute("""
    ATTACH 'ducklake:postgres:dbname=ducklake_catalog' AS ducklake_catalog
        (DATA_PATH 'abfs://data');

    USE ducklake_catalog;
""")

con.execute("""
    -- Create Tesla cars table with sample data
    CREATE TABLE tesla_cars AS
    SELECT * FROM VALUES
        ('Model S', 2024, 'Plaid', 405, 1020, 149990, 'Red Multi-Coat', 'Black', TRUE, TRUE),
        ('Model 3', 2024, 'Performance', 315, 510, 52990, 'Pearl White', 'Black', TRUE, FALSE),
        ('Model 3', 2023, 'Long Range', 358, 346, 47990, 'Midnight Silver Metallic', 'White', FALSE, FALSE),
        ('Model X', 2024, 'Plaid', 348, 1020, 109990, 'Pearl White', 'Black', TRUE, TRUE),
        ('Model Y', 2024, 'Performance', 303, 456, 54990, 'Deep Blue Metallic', 'Black', TRUE, FALSE),
        ('Model Y', 2023, 'Long Range', 330, 346, 50990, 'Red Multi-Coat', 'White', FALSE, FALSE),
        ('Model S', 2023, 'Dual Motor', 405, 670, 89990, 'Midnight Silver Metallic', 'Cream', FALSE, TRUE),
        ('Cybertruck', 2024, 'Tri Motor', 340, 845, 99990, 'Stainless Steel', 'Black', TRUE, TRUE),
        ('Model 3', 2024, 'RWD', 272, 283, 38990, 'Pearl White', 'Black', FALSE, FALSE),
        ('Model X', 2023, 'Dual Motor', 348, 670, 98990, 'Deep Blue Metallic', 'Cream', FALSE, TRUE)
    AS t(model, year, trim, range_miles, horsepower, price_usd, exterior_color, interior_color, autopilot_included, premium_connectivity);
""")
```

**Comment by DrewScatterday** [comment]
> > Fsspec works from the python side and supports the data warehouse etl workflow, but with new users connecting with ducklake, we'll want this supported in the extension and via the duckdb web or dbeaver ide or similar connection methods. This issue just gained a lot of demand from the community.
> 
> [@jordandakota](https://github.com/jordandakota) Would you be able to share how you were able to get ducklake working with ADLS Gens/Blob storage using fsspec? I did what you mentioned above, and now getting a different error when I try to create a table in ducklake using an already existing container `data` in my ADLS Gen2 storage account:
> 
> `IO Error: Failed to create directory "abfs://data/": No such file or directory`
> 
> Code:
> 
> import os
> import duckdb
> import adlfs
> from fsspec import filesystem
> 
> con = duckdb.connect()
> duckdb.register_filesystem(filesystem("abfs", connection_string=os.getenv("AZURE_CONNECTION_STRING")))
> 
> con.execute("""
>     INSTALL ducklake;
>     INSTALL postgres;
> 
>     LOAD ducklake;
>     LOAD postgres;
> """)
> 
> con.execute(f"""
>     CREATE SECRET (
>         TYPE postgres,
>         HOST '{os.environ["POSTGRES_SERVER"]}',
>         PORT 5432,
>         DATABASE postgres,
>         USER '{os.environ["POSTGRES_USER"]}',
>         PASSWORD '{os.environ["POSTGRES_PASS"]}'
>     );
> """)
> 
> # Setup data path to container "data" that already exists in my ADLS Gen2 Storage Account
> con.execute("""
>     ATTACH 'ducklake:postgres:dbname=ducklake_catalog' AS ducklake_catalog
>         (DATA_PATH 'abfs://data');
> 
>     USE ducklake_catalog;
> """)
> 
> con.execute("""
>     -- Create Tesla cars table with sample data
>     CREATE TABLE tesla_cars AS
>     SELECT * FROM VALUES
>         ('Model S', 2024, 'Plaid', 405, 1020, 149990, 'Red Multi-Coat', 'Black', TRUE, TRUE),
>         ('Model 3', 2024, 'Performance', 315, 510, 52990, 'Pearl White', 'Black', TRUE, FALSE),
>         ('Model 3', 2023, 'Long Range', 358, 346, 47990, 'Midnight Silver Metallic', 'White', FALSE, FALSE),
>         ('Model X', 2024, 'Plaid', 348, 1020, 109990, 'Pearl White', 'Black', TRUE, TRUE),
>         ('Model Y', 2024, 'Performance', 303, 456, 54990, 'Deep Blue Metallic', 'Black', TRUE, FALSE),
>         ('Model Y', 2023, 'Long Range', 330, 346, 50990, 'Red Multi-Coat', 'White', FALSE, FALSE),
>         ('Model S', 2023, 'Dual Motor', 405, 670, 89990, 'Midnight Silver Metallic', 'Cream', FALSE, TRUE),
>         ('Cybertruck', 2024, 'Tri Motor', 340, 845, 99990, 'Stainless Steel', 'Black', TRUE, TRUE),
>         ('Model 3', 2024, 'RWD', 272, 283, 38990, 'Pearl White', 'Black', FALSE, FALSE),
>         ('Model X', 2023, 'Dual Motor', 348, 670, 98990, 'Deep Blue Metallic', 'Cream', FALSE, TRUE)
>     AS t(model, year, trim, range_miles, horsepower, price_usd, exterior_color, interior_color, autopilot_included, premium_connectivity);
> """)

Do ‘con.register_filesystem’ instead of ‘duckdb.register_filesytem’. Took me way too long to track that down the first time it happened to me haha

**Comment by crazy-treyn** [comment]
Wow... I should have caught that from when you first shared it, sorry about that. Thank you!!

**Comment by tang2087** [comment]
> Do `con.register_filesystem` instead of `duckdb.register_filesytem`. Took me way too long to track that down the first time it happened to me haha

Thank you!
After install the following two python packages and register the file system in conn, I am able to get ducklake running with Azure Blob Storage:
```
fsspec
adlfs
```

In case anyone only installed fsspec,  you will see the following error:

ImportError: Install adlfs to access Azure Datalake Gen2 and Azure Blob Storage


Anyone knows when the azure extension will have this enabled by default to make it easier?

**Comment by njust** [comment]
The docs [here](https://ducklake.select/docs/stable/duckdb/usage/choosing_storage) are misleading: 

"The DuckDB ducklake extension can work with any file system backend that DuckDB supports. This currently includes: 
...
[Azure Blob Store](https://duckdb.org/docs/stable/core_extensions/azure.html)
"

The docs should make it clear that only read operations on azure blob storage are currently supported, unless the python sdk and the fsspec / adlfs extensions are used.

Related issue https://github.com/duckdb/duckdb-azure/issues/50

**Comment by ramkumarkb** [comment]
I tried the `fsspec` + `adlfs` work around with DuckDB `1.3.1` and DuckLake `0.2.0` - and it did not work.

I got the same `Not Implemented` error.

**Comment by mrjsj** [comment]
> I tried the `fsspec` + `adlfs` work around with DuckDB `1.3.1` and DuckLake `0.2.0` - and it did not work.
> 
> I got the same `Not Implemented` error. 

Can you share the code you tried?
Make sure you're registering the file system on the connection rather than duckdb instance, and that you're using abfs - not abfss.

**Comment by MiguelElGallo** [comment]
In my opinion, now that ducklake is here, the need to be able to write to Azure without (FSSPEC) which is only python, is getting stronger. 

I mean if someone can write the steps needs I will we happy to help with the coding.

**Comment by jordandakota** [comment]
> In my opinion, now that ducklake is here, the need to be able to write to Azure without (FSSPEC) which is only python, is getting stronger. 
> 
> I mean if someone can write the step needed I will we happy to help with the coding.
> 

I think it's waiting on a major refactor by Mytherin. https://github.com/duckdb/duckdb/pull/11259

**Comment by natemcintosh** [comment]
Has anyone been able to access a plain DuckDB database in `read+write` mode that lives in Azure Blob? I'm on using DuckDB 1.4.4 and azure extension `7e1ac33`. I can access a DuckDB database in read mode, but when I attempt to access in `read+write` mode, I get:
```sql
attach 'az://<container-name>/test.db' (READ_WRITE);
-- Not implemented Error:
-- Unsupported: cannot open an Azure file in read+write mode
```
I've also tried using `abfs://` instead of `az://` with the same results.

**Comment by benfleis** [comment]
Attaching to remote DuckDB databases is generally unsupported, in Azure, S3, etc.

Some people have used a work-around with e.g. FUSE and other similar remote filesystem mappers.

However, we recommend **strongly against** doing this, as DuckDB cannot guarantee data integrity when doing so.

> attach 'az://<container-name>/test.db' (READ_WRITE);
> -- Not implemented Error:
> -- Unsupported: cannot open an Azure file in read+write mode
> I've also tried using `abfs://` instead of `az://` with the same results.
